### PR TITLE
chore(architecture): add per-batch timeout support to Batcher

### PR DESCRIPTION
**Pull request by tobz**
This PR primarily focuses on changing `vector_core::stream::batcher::Batcher` from a "flush all batches at the same time" to an a per-batch timeout model.  Instead of flushing all batches at the same exact time, which can lead to bursts of activity that apply backpressure to other components in a pipeline, we apply timeouts at the batch level.  This means that the timeout for a batch starts when the batch is created.  As batches are typically constructed due to partitioning of the output -- partitioning based on the value of a field in the event, for example -- they tend to be spread out over time.

Overall, this leads to per-batch timeouts being important for spreading out the load over time, and providing _consistent_ throughput/latency to pipelines.

Additionally, this PR also removes the constraint on `Partitioner` of returning an `Option<Self::Key>`, instead switching purely to `Self::Key`, which allows for partitioner implementations to decide whether or not they must support fallibility.

### CODE REVIEW

File: lib/vector-core/Cargo.toml [hunk 1]
```diff
@@ -10,6 +10,7 @@ async-graphql = { version = "=2.6.4", default-features = false, optional = true
 async-trait = { version = "0.1", default-features = false }
 atomig = { version = "0.3.1", features = ["derive", "serde"] }
 buffers = { path = "buffers", default-features = false }
+bimap = { version = "0.6.1", default-features = false, features = ["std"] }
```
Comment (blt) [inline lib/vector-core/Cargo.toml:13]: I think we can drop this now?

Comment (tobz) [reply lib/vector-core/Cargo.toml:13]: Woops, yeah, good call.

File: lib/vector-core/src/stream/batcher.rs [hunk 2]
```diff
@@ -529,4 +647,67 @@ mod test {
             }
         }
     }
+
+    // WIP: this currently won't run because we're pointed to a git checkout of tokio-util, which
+    // uses a path-based dependency on tokio; since we use versioned tokio, this creates a mismatch
+    // between the thread-local variables/globals that the timer code uses to find the current
+    // runtime, and as such, `DelayQueue` will never find the right runtime
+    //
+    // we need to get tokio-util 0.6.8 released
```
Comment (blt) [inline lib/vector-core/src/stream/batcher.rs:656]: Comment is out of date now?

**Review by blt — APPROVED** [review]
A few small comments to resolve but I'm +1 on this change.

File: lib/vector-core/src/stream/batcher.rs (hunk 2 above)
Comment (tobz) [reply lib/vector-core/src/stream/batcher.rs:656]: Correctamundo!

Comment (tobz) [reply lib/vector-core/src/stream/batcher.rs]: Addressed in a70f70e8f.

File: lib/vector-core/Cargo.toml (hunk 1 above)
Comment (tobz) [reply lib/vector-core/Cargo.toml]: Addressed in a70f70e8f

**Comment by blt** [comment]
Resolved #8973

**Pull request merged**